### PR TITLE
Issue #4352: Added ability to skip CI executions base on commit files i…

### PR DESCRIPTION
**Pull request by Gaurav-Punjabi**
Fixes issue #4352

### CODE REVIEW

**Comment by Gaurav-Punjabi** [comment]
@romani Please review.

File: .ci/run-link-check-plugin.sh [hunk 1]
```diff
@@ -3,19 +3,25 @@
...
-# "grep ... | cat" is required command is running in "set -e" mode and
-# grep could return exit code 1 if nothing is matching
-grep externalLink target/site/linkcheck.html | cat
-echo "------------ grep of linkcheck.html--END"
-RESULT=$(grep externalLink target/site/linkcheck.html | grep -v 'Read timed out' | wc -l)
-echo 'Exit code:'$RESULT
-if [[ $RESULT != 0 ]]; then false; fi
+# To skip build if minor files or other CI files area changed
+SKIP_FILES=".github|appveyor.yml|circle.yml|distelli-manifest.yml|fast-forward-merge.sh|LICENSE|\
+            LICENSE.apache20|README.md|release.sh|RIGHTS.antlr|shippable.yml|wercker.yml"
```
Comment (romani) [inline .ci/run-link-check-plugin.sh]: please add this code to https://github.com/checkstyle/checkstyle/blob/master/codeship-steps.yml 
we need to skip only when we are in CI mode. This shell file can be executed at local and it should not skip anything.

Comment (Gaurav-Punjabi) [reply .ci/run-link-check-plugin.sh]: Alright, got it 👍

Comment (Gaurav-Punjabi) [reply .ci/run-link-check-plugin.sh]: @romani Done.

File: codeship-steps.yml [hunk 2]
```diff
@@ -7,4 +7,5 @@
     #-  command: ./.ci/travis/travis.sh site
     #-  command: ./.ci/travis/travis.sh nondex
     -  command: ls -la
-    -  command: .ci/run-link-check-plugin.sh
+    -  command: SKIP_FILES=".github|appveyor.yml|circle.yml|distelli-manifest.yml|fast-forward-merge.sh|LICENSE|LICENSE.apache20|README.md|release.sh|RIGHTS.antlr|shippable.yml|wercker.yml"
+    -  command: if [[ $(git diff --name-only HEAD HEAD~1 | grep -vE "$SKIP_FILES" | cat | wc -c) > 0 ]]; then .ci/run-link-check-plugin.sh; fi
```
Comment (romani) [inline codeship-steps.yml]: please wrap long lines to be less then 100 symbols in length.

Comment (Gaurav-Punjabi) [reply codeship-steps.yml]: Sure

Comment (Gaurav-Punjabi) [reply codeship-steps.yml]: Done.

File: codeship-steps.yml [hunk 3]
```diff
@@ -6,5 +6,10 @@
   steps:
     #-  command: ./.ci/travis/travis.sh site
     #-  command: ./.ci/travis/travis.sh nondex
-    -  command: ls -la
-    -  command: .ci/run-link-check-plugin.sh
+    - command: ls -la
+    - command: SKIP_FILES=
+        ".github|appveyor.yml|circle.yml|distelli-manifest.yml|fast-forward-merge.sh|LICENSE|
```
Comment (romani) [inline codeship-steps.yml:11]: please use this form  https://github.com/checkstyle/checkstyle/blob/72a5294619a788a9e1c4e36e4df8177301ef1f9b/shippable.yml#L42

please make single `command` node with `- |` as it is done at https://github.com/checkstyle/checkstyle/blob/72a5294619a788a9e1c4e36e4df8177301ef1f9b/shippable.yml#L32

Comment (Gaurav-Punjabi) [reply codeship-steps.yml:11]: Alright 👍

Comment (Gaurav-Punjabi) [reply codeship-steps.yml]: Made the changes. Please reveiw

File: codeship-steps.yml [hunk 4]
```diff
@@ -6,5 +6,11 @@
...
-    -  command: ls -la
-    -  command: .ci/run-link-check-plugin.sh
+    - command: |
+        ls -la
+        SKIP_FILES1=".github|distelli-manifest.yml|fast-forward-merge.sh"
+        SKIP_FILES2="LICENSE|LICENSE.apache20|README.md|release.sh|RIGHTS.antlr"
+        SKIP_FILES3="shippable.yml|wercker.yml|appveyor.yml|circle.yml"
+        SKIP_FILES="$SKIP_FILES1$SKIP_FILES2$SKIP_FILES3"
+        if [[ $(git diff --name-only HEAD HEAD~1 | grep -vE "$SKIP_FILES" | cat | wc -c) > 0 ]];
+        then .ci/run-link-check-plugin.sh; fi
```
Comment (romani) [inline codeship-steps.yml:16]: Please move `then` and `fi` to separate lines, we respect indentation rules even in bash

Comment (Gaurav-Punjabi) [reply codeship-steps.yml]: Done 👍

File: codeship-steps.yml [hunk 5]
```diff
@@ -6,5 +6,13 @@
...
     #-  command: ./.ci/travis/travis.sh site
     #-  command: ./.ci/travis/travis.sh nondex
-    -  command: ls -la
-    -  command: .ci/run-link-check-plugin.sh
+    - command: |
+        ls -la
+        SKIP_FILES1=".github|distelli-manifest.yml|fast-forward-merge.sh"
+        SKIP_FILES2="LICENSE|LICENSE.apache20|README.md|release.sh|RIGHTS.antlr"
+        SKIP_FILES3="shippable.yml|wercker.yml|appveyor.yml|circle.yml"
+        SKIP_FILES="$SKIP_FILES1$SKIP_FILES2$SKIP_FILES3"
```
Comment (pbludov) [inline codeship-steps.yml:14]: The rightmost item should be separated from the leftmost on the next line:
SKIP_FILES="$SKIP_FILES1`|`$SKIP_FILES2`|`$SKIP_FILES3"
 OR
SKIP_FILES1=".github`|`distelli-manifest.yml`|`fast-forward-merge.sh`|`" (and so on)

Comment (pbludov) [inline codeship-steps.yml:14]: The rightmost item should be separated from the leftmost on the next line:
SKIP_FILES="$SKIP_FILES1|$SKIP_FILES2|$SKIP_FILES3"
OR
SKIP_FILES1=".github|distelli-manifest.yml|fast-forward-merge.sh|" (and so on)

Comment (Gaurav-Punjabi) [reply codeship-steps.yml]: Done. Fixed it 👍

File: codeship-steps.yml [hunk 6]
```diff
@@ -6,5 +6,13 @@
...
     #-  command: ./.ci/travis/travis.sh nondex
-    -  command: ls -la
-    -  command: .ci/run-link-check-plugin.sh
+    - command: |
+        ls -la
+        SKIP_FILES1=".github|distelli-manifest.yml|fast-forward-merge.sh"
+        SKIP_FILES2="LICENSE|LICENSE.apache20|README.md|release.sh|RIGHTS.antlr"
+        SKIP_FILES3="shippable.yml|wercker.yml|appveyor.yml|circle.yml"
+        SKIP_FILES="$SKIP_FILES1|$SKIP_FILES2|$SKIP_FILES3"
+        if [[ $(git diff --name-only HEAD HEAD~1 | grep -vE "$SKIP_FILES" | cat | wc -c) > 0 ]];
```
Comment (rnveach) [inline codeship-steps.yml:15]: @romani Is this safe to do when travis and other CIs dropped this and forced to run all the time?
https://github.com/checkstyle/checkstyle/blob/master/.travis.yml#L326

Comment (romani) [reply codeship-steps.yml:15]: it is not safe, but till it present in other CI, lets keep it here too.
Removal of all is another issue, as far as I remember we did not decided on it yet.

**Pull request merged**